### PR TITLE
Feature dtcenter/METplus-Internal#19 log user ID and dtcenter/METplus-Internal#21 signal handling

### PR DESCRIPTION
**Pull request by hsoh-u**
## Expected Differences ##

This is a prototype for the design review for login user ID and adding signal handlers at the centralized location. Note: only ascii2nc was modified to demonstrate the design.
- Common implementation at main.h & main.cc
  - "int main(...)" method is implemented at basic/vx_log/main.cc
  - pre-processing
    - Log message with when, who, which MET apps
    - Install signal handlers
  - calls met_main()
  - post-processing
    - uninstall signal handlers
    - Log message with when, who, which MET apps
- each MET tools (for example, ascii2nc)
  - "int main(...)" method should be renamed to ""int met_main(...)" method 
  - "const char *get_tool_name()" should be implemented
- exit() should be modified to call the post_processing or should create a custom signal

- [ ] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes or No]**</br>
If **yes**, please describe:</br>

- [ ] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[Yes or No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [] Describe testing already performed for these changes:</br>

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[No]**

- [x] Do these changes include sufficient testing updates? **[No]**

- [x] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Linked issues** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: met/src/basic/vx_log/main.cc [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+void do_pre_process(uid_t user_id, const char *tool_name) {
+   cout << "  Started " << tool_name << " by user " << user_id << " at " << get_current_time() << "\n";
+   cout << "  Do signal handling here and other pre-processing\n";
+}
+
+
+////////////////////////////////////////////////////////////////////////
+
+void do_post_preocess(uid_t user_id, const char *tool_name) {
```
Comment (georgemccabe) [inline met/src/basic/vx_log/main.cc]:
```suggestion
void do_post_process(uid_t user_id, const char *tool_name) {
```

File: met/src/basic/vx_log/main.cc [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+int main(int argc, char *argv[]) {
+
+   uid_t user_id = get_user_id();
+   const char * tool_name = get_tool_name();
+
+   do_pre_process(user_id, tool_name);
+
+   int return_code = met_main(argc, argv);
+
+   do_post_preocess(user_id, tool_name);
```
Comment (georgemccabe) [inline met/src/basic/vx_log/main.cc]:
```suggestion
   do_post_process(user_id, tool_name);
```

File: met/src/basic/vx_log/main.cc [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+   cout << "  Started " << tool_name << " by user " << user_id << " at " << get_current_time() << "\n";
+   cout << "  Do signal handling here and other pre-processing\n";
+}
+
+
+////////////////////////////////////////////////////////////////////////
+
+void do_post_preocess(uid_t user_id, const char *tool_name) {
+   cout << "  Do post-processing\n";
+   cout << "  Done  " << tool_name << " by user " << user_id << " at " << get_current_time() << "\n";
```
Comment (georgemccabe) [inline met/src/basic/vx_log/main.cc]: Consider changing "Done" to "Finished" or "Completed"

**Review by georgemccabe — APPROVED** [review]
These changes look like a good approach to adding the desired information to the run of each MET tool. The changes needed for each other tool would be minimal with this approach. I made a few suggested edits inline. There are a few commented lines of code that could be removed as well. I approve.

File: src/basic/vx_log/main.cc [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+   do_pre_process(user_id, tool_name);
+
+   int return_code = met_main(argc, argv);
+
+   do_post_preocess(user_id, tool_name);
+
+   return return_code;
+
+}
```
Comment (JohnHalleyGotway) [inline src/basic/vx_log/main.cc]:
```suggestion
}

```

File: src/basic/vx_log/main.cc [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+// ** Research Applications Lab (RAL)
+// ** P.O.Box 3000, Boulder, Colorado, 80307-3000, USA
+// *=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*=*
+
+////////////////////////////////////////////////////////////////////////
+//
+//   Filename:   main.cc
+//
+//   Description:
+//      The main() should be inplemented here and the exisyting main() should be renamed to met_main.
```
Comment (JohnHalleyGotway) [inline src/basic/vx_log/main.cc]:
```suggestion
//      The main() should be implemented here and the existing main()
//      should be renamed to met_main().
//
////////////////////////////////////////////////////////////////////////
```

**Review by JohnHalleyGotway — CHANGES_REQUESTED** [review]
Howard, I get the general idea here, creating a wrapper for the existing main() function of each tool. And it seems like a promising approach. One issue I see is writing the timestamps to cout instead of via the logger. Is there a way to route the messages through mlog instead so that they can be save to the generated log file?

**Comment by hsoh-u** [comment]
Typos were corrected.
More code changes:
- main() at main.cc
  - use mlog
  - calls set_new_handler(oom)
    - this causes link problems, so Makefile.am was modified
- The source codes under src/tools were modified
  - main is renamed to met_main
  - get_tool_name() is added
  - set_new_handler(oom) is removed because it is called at the main.cc.
  - Note: src/basic/enum_to_string/enum_to_string was not updated (let me know this should be updated, too)
 
To do list:
- move initalize() and process_command_line to main() at main.cc
  - not updated yet because this causes another link issues
- log event (start/finished) to the centralized place

**Review by JohnHalleyGotway — APPROVED** [review]
@hsoh-u there is a single main() function you missed updating:
   [src/tools/other/wwmca_tool/wwmca_regrid.cc:int main(int argc, char * argv [])](https://github.com/dtcenter/MET/blob/d2e0a7502e869c80ac081a43cfa53b116db0c690/src/tools/other/wwmca_tool/wwmca_regrid.cc#L83)

After updating that one, please proceed with the merge.
I see via GHA that the code still compiles and produces no unexpected diffs.
We should include this change in the beta2 release tomorrow so that we have plenty of time to test it out and make sure it doesn't cause any unintended compilation/runtime consequences.

I approve, subject to updating wwmca_regrid.cc.

Thanks for all your detailed work on this!

**Comment by JohnHalleyGotway** [comment]
Note that this issue also resolved dtcenter/METplus-Internal#18 since it logs the beginning/ending time of each run.

**Pull request merged**